application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/LicensesNotificationsUIX.xml

```diff
@@ -38,39 +38,72 @@
+    &lt;p&gt;$services.localization.render($translation)&lt;/p&gt;
+    &lt;ul&gt;
+      &lt;li&gt;$stringtool.join($extensionNames, '&lt;/li&gt;&lt;li&gt;')&lt;/li&gt;
+      #if ($showMore)
+        &lt;li&gt;...&lt;/li&gt;
+      #end
+    &lt;/ul&gt;
+  &lt;/$missingLicensesWrapperTagName&gt;
+#end
+{{/velocity}}
+
 {{velocity}}
 #if ($hasAdmin &amp;&amp; $hasProgramming)
   #getVisiblePaidExtensions
   ## The same extension could be installed on multiple namespaces (wikis) with different versions.
-  #set ($extensionNames = $collectiontool.sortedSet)
+  #set ($invalidExtensionNames = $collectiontool.sortedSet)
+  #set ($expiringExtensionNames = $collectiontool.sortedSet)
   #set ($limit = 3)
-  #set ($showMore = false)
+  #set ($showMoreInvalid = false)
+  #set ($showMoreExpiring = false)
+  #set ($foundExpiring = false)
+  #set ($foundInvalid = false)
   #foreach ($extension in $visiblePaidExtensions)
-    #if (!$services.licensing.licensor.hasLicensureForExtension($extension.id))
-      #if ($extensionNames.size() &lt; $limit)
-        #set ($discard = $extensionNames.add($escapetool.xml($extension.name)))
+    #if (!$showMoreInvalid &amp;&amp; !$services.licensing.licensor.hasLicensureForExtension($extension.id))
+      #set ($foundInvalid = true)
+      #if ($invalidExtensionNames.size() &lt; $limit)
+        #set ($discard = $invalidExtensionNames.add($escapetool.xml($extension.name)))
+      #else
+        #set ($showMoreInvalid = true)
+      #end
+    #elseif (!$showMoreExpiring &amp;&amp; $services.licensing.licensor.isLicenseExpiring($extension.id))
+      #set ($foundExpiring = true)
+      #if ($expiringExtensionNames.size() &lt; $limit)
+        #set ($discard = $expiringExtensionNames.add($escapetool.xml($extension.name)))
       #else
-        #set ($showMore = true)
-        #break
+        #set ($showMoreExpiring = true)
       #end
     #end
+    #if ($showMoreExpiring &amp;&amp; $showMoreInvalid)
+      #break
+    #end
   #end
-  #if ($showMore || $extensionNames.size() &gt; 0)
-    #set ($discard = $xwiki.ssx.use('Licenses.Code.LicensesNotificationsUIX'))
+  #if ($foundInvalid || $foundExpiring)
     #if (!$missingLicensesWrapperTagName)
       #set ($missingLicensesWrapperTagName = 'div')
     #end
     {{html clean=false}}
-    &lt;$missingLicensesWrapperTagName class="notifications-missingLicenses errormessage"&gt;
-      &lt;p&gt;$services.localization.render('licensor.missingLicenses', ["&lt;a href='#getLicensesAdminSectionURL'&gt;",
-        '&lt;/a&gt;'])&lt;/p&gt;
-      &lt;ul&gt;
-        &lt;li&gt;$stringtool.join($extensionNames, '&lt;/li&gt;&lt;li&gt;')&lt;/li&gt;
-        #if ($showMore)
-          &lt;li&gt;...&lt;/li&gt;
-        #end
-      &lt;/ul&gt;
+    &lt;$missingLicensesWrapperTagName class="notifications-missingLicenses"&gt;
+      &lt;p&gt;$services.localization.render('licensor.licensesSection', ["&lt;a href='#getLicensesAdminSectionURL'&gt;",
```

Comment:
Was there a reason to move the license section button outside the error / warning bubble? To me right now it's harder to be seen as part of the notification, so can be missed by users

Also, isn't the link escaped in here by localization render? Would be better to use something like https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-ui/src/main/resources/CKEditor/VelocityMacros.xml#L229